bumble/profiles/hap.py

```diff
@@ -0,0 +1,556 @@
+    '''See Hearing Access Service 3.1. Hearing Aid Features.'''
+
+    PRESET_SYNCHRONIZATION_IS_NOT_SUPPORTED = 0b0
+    PRESET_SYNCHRONIZATION_IS_SUPPORTED = 0b1
+
+
+class IndependentPresets(OpenIntEnum):
+    '''See Hearing Access Service 3.1. Hearing Aid Features.'''
+
+    IDENTICAL_PRESET_RECORD = 0b0
+    DIFFERENT_PRESET_RECORD = 0b1
+
+
+class DynamicPresets(OpenIntEnum):
+    '''See Hearing Access Service 3.1. Hearing Aid Features.'''
+
+    PRESET_RECORDS_DOES_NOT_CHANGE = 0b0
+    PRESET_RECORDS_MAY_CHANGE = 0b1
+
+
+class WritablePresetsSupport(OpenIntEnum):
+    '''See Hearing Access Service 3.1. Hearing Aid Features.'''
+
+    WRITABLE_PRESET_RECORDS_NOT_SUPPORTED = 0b0
+    WRITABLE_PRESET_RECORDS_SUPPORTED = 0b1
+
+
+class HearingAidPresetControlPointOpcode(OpenIntEnum):
+    '''See Hearing Access Service 3.3.1 Hearing Aid Preset Control Point operation requirements.'''
+
+    # fmt: off
+    READ_PRESETS_REQUEST                     = 0x01
+    READ_PRESET_RESPONSE                     = 0x02
+    PRESET_CHANGED                           = 0x03
+    WRITE_PRESET_NAME                        = 0x04
+    SET_ACTIVE_PRESET                        = 0x05
+    SET_NEXT_PRESET                          = 0x06
+    SET_PREVIOUS_PRESET                      = 0x07
+    SET_ACTIVE_PRESET_SYNCHRONIZED_LOCALLY   = 0x08
+    SET_NEXT_PRESET_SYNCHRONIZED_LOCALLY     = 0x09
+    SET_PREVIOUS_PRESET_SYNCHRONIZED_LOCALLY = 0x0A
+
+
+class PresetChangedOperation:
+    '''See Hearing Access Service 3.2.2.2. Preset Changed operation.'''
+
+    class ChangeId(OpenIntEnum):
+        # fmt: off
+        GENERIC_UPDATE            = 0x00
+        PRESET_RECORD_DELETED     = 0x01
+        PRESET_RECORD_AVAILABLE   = 0x02
+        PRESET_RECORD_UNAVAILABLE = 0x03
+
+    class Generic:
+        prev_index: int
+        preset_record: PresetRecord
+
+        def __init__(self, idx, preset):
+            self.prev_index = idx
+            self.preset_record = preset
+
+        def to_bytes(self) -> bytes:
+            return bytes([self.prev_index]) + self.preset_record.to_bytes()
+
+    def __init__(self, id: ChangeId, params: Generic):
+        self.change_id = id
+        self.additional_parameters = params
+
+    change_id: ChangeId
+    additional_parameters: Generic | int
+
+    def additional_parameters_to_bytes(self) -> bytes:
+        if isinstance(self.additional_parameters, PresetChangedOperation.Generic):
+            return self.additional_parameters.to_bytes()
+        return bytes([self.additional_parameters])
+
+    def to_bytes(self, is_last: bool) -> bytes:
+        # TODO does the bytes concatenate correctly ?
+        return (
+            bytes(
+                [
+                    HearingAidPresetControlPointOpcode.PRESET_CHANGED,
+                    self.change_id,
+                    is_last,
+                ]
+            )
+            + self.additional_parameters_to_bytes()
+        )
+
+    @staticmethod
```

Comment:
classmethod is preferred over staticmethod

Reply:
classmethod is taking a class type as parameter that will be unused

Why should we prefer it over static ?

Reply:
https://google.github.io/styleguide/pyguide.html#217-function-and-method-decorators

Reply:
Ok, sounds like I should avoid classmethod too.
I inlined it in the only place that was using it as this is just used to sort once and doesn't really need to be expose at module level